core/src/test/java/com/linecorp/armeria/server/auth/AuthorizerTest.java

```diff
@@ -0,0 +1,168 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.auth;
+
+import static com.spotify.futures.CompletableFutures.exceptionallyCompletedFuture;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+import javax.annotation.Nullable;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.channel.DefaultEventLoop;
+import io.netty.channel.EventLoop;
+
+public class AuthorizerTest {
+
+    @Nullable
+    private static EventLoop eventLoop;
+    @Nullable
+    private static ServiceRequestContext serviceCtx;
+
+    @BeforeClass
+    public static void setEventLoop() {
+        eventLoop = new DefaultEventLoop();
+        serviceCtx = mock(ServiceRequestContext.class);
+        when(serviceCtx.contextAwareEventLoop()).thenReturn(eventLoop);
+    }
+
+    @AfterClass
+    public static void clearEventLoop() {
+        serviceCtx = null;
+        if (eventLoop != null) {
+            eventLoop.shutdownGracefully();
+            eventLoop = null;
+        }
+    }
+
+    @Test
+    public void orElseFirst() {
+        final Authorizer<String> a = newMock();
+        when(a.authorize(any(), any())).thenReturn(completedFuture(true));
+        final Authorizer<String> b = newMock();
+
+        final Boolean result = a.orElse(b).authorize(serviceCtx, "data")
+                                .toCompletableFuture().join();
+        assertThat(result).isTrue();
+        verify(a, times(1)).authorize(serviceCtx, "data");
+        verify(b, never()).authorize(any(), any());
+    }
+
+    /**
+     * When the first {@link Authorizer} raises an exception, the second {@link Authorizer} shouldn't be
+     * invoked.
+     */
+    @Test
+    public void orElseFirstException() {
+        final Exception expected = new Exception();
+        final Authorizer<String> a = newMock();
+        when(a.authorize(any(), any())).thenReturn(exceptionallyCompletedFuture(expected));
+        final Authorizer<String> b = newMock();
+
+        assertThatThrownBy(() -> a.orElse(b).authorize(serviceCtx, "data").toCompletableFuture().join())
+                .isInstanceOf(CompletionException.class)
+                .hasCause(expected);
+        verify(a, times(1)).authorize(serviceCtx, "data");
+        verify(b, never()).authorize(any(), any());
```

Comment:
You can use `verifyZeroInteractions` too if you want (maybe it's semantically simpler to say "Authorizer was never used" instead of "Authorizer's authorize method was never called). But either way's fine.

Reply:
+1. Fixed :-)